breacharbiter.go

```diff
@@ -675,122 +554,74 @@ secondLevelCheck:
-// only after the breaching transaction receives a necessary number of
-// confirmations.
-func (b *breachArbiter) breachObserver(
-	chainEvents *contractcourt.ChainEventSubscription,
-	settleSignal chan struct{}) {
-
-	defer func() {
-		b.wg.Done()
-		chainEvents.Cancel()
-	}()
-
-	chanPoint := chainEvents.ChanPoint
-
-	brarLog.Debugf("Breach observer for ChannelPoint(%v) started ",
-		chanPoint)
-
-	select {
-	// A read from this channel indicates that the contract has been
-	// settled cooperatively so we exit as our duties are no longer needed.
-	case <-settleSignal:
-		return
+// handleBreachHandoff handles a new breach event, by writing it to disk, then
+// notifies the breachArbiter contract observer goroutine that a channel's
+// contract has been breached by the prior counterparty. Once notified the
+// breachArbiter will attempt to sweep ALL funds within the channel using the
+// information provided within the BreachRetribution generated due to the
+// breach of channel contract. The funds will be swept only after the breaching
+// transaction receives a necessary number of confirmations.
+//
+// NOTE: This MUST be run as a goroutine.
+func (b *breachArbiter) handleBreachHandoff(breachEvent *ContractBreachEvent) {
+	defer b.wg.Done()
 
-	// The channel has been closed cooperatively, so we're done here.
-	case <-chainEvents.CooperativeClosure:
-		// Launch a goroutine to cancel out this contract within the
-		// breachArbiter's main goroutine.
-		b.wg.Add(1)
-		go func() {
-			defer b.wg.Done()
+	chanPoint := breachEvent.ChanPoint
+	brarLog.Debugf("Handling breach handoff for ChannelPoint(%v)",
+		chanPoint)
 
-			select {
-			case b.settledContracts <- chanPoint:
-			case <-b.quit:
-			}
-		}()
+	// A read from this channel indicates that a channel breach has been
+	// detected! So we notify the main coordination goroutine with the
+	// information needed to bring the counterparty to justice.
+	breachInfo := breachEvent.BreachRetribution
+	brarLog.Warnf("REVOKED STATE #%v FOR ChannelPoint(%v) "+
+		"broadcast, REMOTE PEER IS DOING SOMETHING "+
+		"SKETCHY!!!", breachInfo.RevokedStateNum,
+		chanPoint)
 
-		b.cfg.CloseLink(&chanPoint, htlcswitch.CloseBreach)
+	// Immediately notify the HTLC switch that this link has been
+	// breached in order to ensure any incoming or outgoing
+	// multi-hop HTLCs aren't sent over this link, nor any other
+	// links associated with this peer.
+	b.cfg.CloseLink(&chanPoint, htlcswitch.CloseBreach)
 
-	// The channel has been closed by a normal means: force closing with
-	// the latest commitment transaction.
-	case <-chainEvents.UnilateralClosure:
-		// Launch a goroutine to cancel out this contract within the
-		// breachArbiter's main goroutine.
-		b.wg.Add(1)
-		go func() {
-			defer b.wg.Done()
+	// TODO(roasbeef): need to handle case of remote broadcast
+	// mid-local initiated state-transition, possible
+	// false-positive?
```

Comment:
Below this line is where we may want to call `IsBreached` to see if this channel has already had an `exactRetribution` task spawned. If it's already been breached, it should be safe to just respond with `nil` and then do nothing, since we should have resumed the `exactRetribution` task on startup

Reply:
Added a check to `IsBreached`, that if successful will ACK and return. Had to introduce a mutex that ensures consistency between the calls to `IsBreached` and `Add`. 

PTAL.